File: RomanArabicNumerals/JavaRomanNumerals/RomanArabic.java
public class RomanArabic {
	private static String[] romanNumbers = 
		{ "I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M" };
	private static int[] arabicValues = 
		{ 1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000 };
	
	public static String arabic2roman (int arabic) throws RomanArabicException {
		if (arabic<=0 || arabic>3999) {
			throw new RomanArabicException();
		}
		String roman = "";
		int pointer = 12;
		while (arabic > 0) {
			if (RomanArabic.arabicValues[pointer]<=arabic) {
				roman += RomanArabic.romanNumbers[pointer];
				arabic-=RomanArabic.arabicValues[pointer];
			}
			else {
				pointer-=1;
			}
		}
		return roman;
	}
	
	public static int roman2arabic (String roman) throws RomanArabicException {
		int pointer=0;
		int arabic=0;
		boolean is_found;
		String current;
		while (pointer<roman.length()) {
			is_found = false;
			if (roman.length()-pointer>1) {
				current = Character.toString(roman.charAt(pointer))+Character.toString(roman.charAt(pointer+1));
				for (int i=0; i<=12; i++) {
					if (current.equals( RomanArabic.romanNumbers[i])) {
						arabic += RomanArabic.arabicValues[i];
						pointer+=2;
						is_found=true;
						break;
					}
				}
				if (is_found) {
					continue;
				}
			}
			current = Character.toString(roman.charAt(pointer));
			for (int i=0; i<=12; i++) {
				if (current.equals( RomanArabic.romanNumbers[i])) {
					arabic+=RomanArabic.arabicValues[i];
					pointer+=1;
					is_found=true;
				}
			}
			if (is_found==false) {
				throw new RomanArabicException();
			}			
		}
		if (roman.equals(RomanArabic.arabic2roman(arabic))) {
			return arabic;
		}
		else {
			throw new RomanArabicException();
		}
	}
}

File: RomanArabic.py
class RomanArabic():

    _romanNumbers = ["I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M"];
    _arabicValues = [1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000];

    def arabic2roman(number,printing=False):
        if (type(number) != int):
            return "{} nie jest liczba calkowita".format(number);
        elif (number>3999 or number<1):
            return "Liczba {} nie miesci sie w przedziale <1, 3999>.".format(number);
        num = number;
        string = "";
        pointer=12;
        while (num>0):
            if (RomanArabic._arabicValues[pointer]<=num):
                string+=RomanArabic._romanNumbers[pointer];
                num-=RomanArabic._arabicValues[pointer];
            else:
                pointer-=1;
        if printing:
            return "{} w systemie rzymskim to: {}".format(number,string);
        else:
            return string;

    def roman2arabic(number):
        if(type(number) != str):
            return "{} nie jest stringiem".format(number);
        pointer = 0;
        value = 0;
        while(pointer<len(number)):
            if (len(number)-pointer>1):
                if(number[pointer]+number[pointer+1] in RomanArabic._romanNumbers):
                    value+=RomanArabic._arabicValues[RomanArabic._romanNumbers.index(number[pointer]+number[pointer+1])];
                    pointer+=2;
                elif (number[pointer] in RomanArabic._romanNumbers):
                    value += RomanArabic._arabicValues[RomanArabic._romanNumbers.index(number[pointer])];
                    pointer += 1;
            elif (number[pointer] in RomanArabic._romanNumbers):
                value+=RomanArabic._arabicValues[RomanArabic._romanNumbers.index(number[pointer])];
                pointer+=1;
            else:
                return "Nie rozpoznano znaku w {}".format(number);
        if (number == RomanArabic.arabic2roman(value)):
            return "{} to {} w systemie arabskim".format(number,value);
        else:
            return"{} nie jest poprawnie zapisana w systemie rzymskim".format(number);


if __name__== "__main__":
    for i in ["MDCVI","CCLVIII", 12, "MMXDCVI","XDCA", "XDDD"]:
        print(RomanArabic.roman2arabic(i));

    print("\n");

    for i in [258, -1, 1999, "abab", 3999, 3.5, 5000, 2646, "xd"]:
        print(RomanArabic.arabic2roman(i,True));

# PwrExercises
